File: benchmarks/10_array_stats/stats.cpp
// Benchmark 10 — Mean + variance of 10 000 000 values  (C++)
// g++ -O3 -march=native -std=c++17 -o stats_cpp stats.cpp && ./stats_cpp
#include <cinttypes>
#include <cstdint>
#include <cstdio>
#include <vector>

int main() {
    const int64_t N = 10000000LL;
    std::vector<int64_t> data((size_t)N);
    for (int64_t i = 0; i < N; i++) data[(size_t)i] = (i * 7 + 13) % 1000;

    int64_t sum = 0;
    for (int64_t i = 0; i < N; i++) sum += data[(size_t)i];
    int64_t mean = sum / N;

    int64_t var_sum = 0;
    for (int64_t i = 0; i < N; i++) {
        int64_t diff = data[(size_t)i] - mean;
        var_sum += diff * diff;
    }
    int64_t variance = var_sum / N;

    std::printf("%" PRId64 "\n%" PRId64 "\n", mean, variance);
}

File: benchmarks/10_array_stats/stats.rs
// Benchmark 10 — Mean + variance of 10 000 000 values  (Rust, sequential)
//
// Build:
//   rustc -C opt-level=3 -C target-cpu=native -o stats_rs stats.rs && ./stats_rs
//
// REVIEWER NOTE: This is a SEQUENTIAL baseline. The vāṇī variant uses two
// `parallel for … reduce` passes (OpenMP-backed). Comparing parallel vs sequential
// measures parallelism strategy, not language quality.
//
// For a fair parallel vs parallel comparison, see stats_rayon.rs (Rayon) and
// stats_omp.c (C + OpenMP). Those should all cluster at ~35-40 ms on 4+ cores.
//
// This sequential file (~65 ms) is kept as the baseline to show the raw
// single-core throughput before parallelism is applied.

fn main() {
    const N: i64 = 10_000_000;
    let data: Vec<i64> = (0..N).map(|i| (i * 7 + 13) % 1000).collect();

    let sum: i64 = data.iter().sum();
    let mean = sum / N;

    let variance: i64 = data.iter().map(|&x| { let d = x - mean; d * d }).sum::<i64>() / N;

    println!("{}", mean);
    println!("{}", variance);
}
